src/gprs.c

```diff
@@ -536,9 +538,171 @@ static void pri_context_signal_settings(struct pri_context *ctx,
 				context_settings_append_ipv6);
 }
 
-static void pri_parse_proxy(struct pri_context *ctx, const char *proxy)
+static void set_default_dns(char **dns_servers)
 {
-	char *scheme, *host, *port, *path;
+	struct in_addr server_in_addr;
+	int i;
+
+	/*
+	 * In this function we modify the global variable _res from resolv.h to
+	 * be able to use a specific DNS server to resolve the MMS proxy/MMSC
+	 * address. This is a hack, although used in some places. Also, the IPv6
+	 * case cannot be handled this way. ldns is a possible option
+	 * (http://www.nlnetlabs.nl/projects/ldns/), but it would introduce a
+	 * new dependence. TODO: search alternatives.
+	 */
+	_res.nscount = 0;
+
+	for (i = 0; dns_servers[i] != NULL; ++i) {
+		if (inet_pton(AF_INET, dns_servers[i], &server_in_addr) != 1)
+			continue;
+
+		_res.nsaddr_list[_res.nscount].sin_addr = server_in_addr;
+		++(_res.nscount);
+
+		/* Size of nsaddr_list array (see resolv.h) */
+		if (_res.nscount == MAXNS)
+			break;
+	}
+}
+
+static gboolean lookup_address(const struct context_settings *settings,
+						const char *proxy,
+						struct sockaddr_in *addr)
+{
+	static gboolean first_time = TRUE;
+	struct addrinfo *result = NULL;
+	struct addrinfo hint;
+	int rc, ret = FALSE;
+
+	DBG("hostname is %s", proxy);
+
+	/*
+	 * We need to initialize the resolv library before changing the DNS
+	 * server. gethostbyname("localhost") does that and at the same time
+	 * does not generate network traffic. Calling directly res_init does
+	 * not really work: that function is called internally again in the
+	 * first call to a resolv function.
+	 */
+	if (first_time) {
+		gethostbyname("localhost");
+		first_time = FALSE;
+	}
+
+	set_default_dns(settings->ipv4->dns);
+
+	memset(&hint, 0, sizeof(hint));
+	hint.ai_socktype = SOCK_STREAM;
+	rc = getaddrinfo(proxy, NULL /* service */, &hint, &result);
```

Comment:
I have been doing some testing, forcing a 2G connection. Lookup times for hosts that exist are always below 2 seconds. For hosts that do not it varies, but only in a few cases I've seen times of more that 3 seconds. I have not seen a defined timeout for getaddrinfo either.

getaddrinfo_a might be a solution to specify a timeout, but it lacks integration with glib. Therefore, we would have to wait in a call to gai_suspend for the lookup to finish, losing the ability to attend other events in ofono meanwhile.